### ISSUE TITLE
Jacoco does not resolve the agent jar from buildscript repositories

### ISSUE THREAD

**Issue by dgeissl** [gradle/gradle#10837]
### Expected Behavior
The jacoco plugin should resolve the jacoco.jar from the buildscript repositories.

### Current Behavior
The plugin actually uses the project repositories to resolve dependencies. These may not contain the `org.jacoco.agent.jar` and fail the build.

### Context
We are using a very limited repository for artifact resolution related to production code which does not contain any jacoco jars. The buildtools may however use a lot more third party libraries. Currently we can not use jacoco cause it is failing our build.

### Steps to Reproduce 
I've attached a small reproducer:
[gradle-jacoco.zip](https://github.com/gradle/gradle/files/3646203/gradle-jacoco.zip) (updated)

This is the result when running the test:
```
$ ./gradlew test
> Task :test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> Could not resolve all files for configuration ':jacocoAgent'.
   > Could not find org.jacoco:org.jacoco.agent:0.8.2.
     Searched in the following locations:
       - file:/C:/Java/Sources/Temp/gradle-jacoco/libs/org.jacoco.agent-0.8.2.jar
       - file:/C:/Java/Sources/Temp/gradle-jacoco/libs/org.jacoco.agent.jar
     Required by:
         project :

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s
2 actionable tasks: 1 executed, 1 up-to-date
```

Maybe these lines must be changed:

they use the `project.getDependencies` api and possibly have to be rewritten to smth. like `project.getBuildscript().getDependencies`?

https://github.com/gradle/gradle/blob/master/subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPlugin.java#L113
https://github.com/gradle/gradle/blob/master/subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPlugin.java#L118
https://github.com/gradle/gradle/blob/master/subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPlugin.java#L130
https://github.com/gradle/gradle/blob/master/subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPlugin.java#L140

### Your Environment
Gradle 5.5.2 or 5.5.1

**Comment by big-guy** [comment]
Thanks @dgeissl - I don't think we want to resolve Jacoco through buildscript dependencies like this.

For your environment, maybe you could get away with restricting an external repository to just the jacoco dependencies using repository matching: https://docs.gradle.org/current/userguide/declaring_repositories.html#sec::matching_repositories_to_dependencies

This would prevent other dependencies from being resolved by the "jacoco only repository".

**Comment by dgeissl** [comment]
Thanks for the feedback @big-guy .
While I'm sure there are better ways to solve it I still wonder what's the difference of jacoco dependencies vs. a plugin that declares lets say `commons-lang` in it's dependency block. 
* Both are needed for the plugin to do it's job
* none of the dependencies are related to the project that the plugin is beeing applied to
but still they should be resolved by different repositories.

Don't get me wrong, I just want to understand the reason to make better decisions in the future.

For your solution, it may restrict the artifacts resolvable through the repository but it still does not prevent the user to resolve jacoco artifacts from the "non-build-code", so extra validation (or exclusions or what ever) are required.

The other problem I did not mention yet is, that we are transitioning a legacy application to gradle and the repository used for project dependencies has some handcrafted metadata (manipulated poms of open source libraries). That's why we also want to keep build dependencies clearly separated from the customized repository as we must not mix them up (with possibly unpredictable results).
That was one feature why we decided to use gradle over maven.
